lms/services/assignment.py

```diff
@@ -56,59 +71,49 @@ def get_copied_from_assignment(self, lti_params) -> Optional[Assignment]:
+        # Get the potentially relevant assignments from the DB
+        assignment = self.get_assignment(tool_consumer_instance_guid, resource_link_id)
+        historical_assignment = None
+        if not assignment:
+            historical_assignment = self.get_copied_from_assignment(lti_params)
 
+        document_url = self._misc_plugin.get_document_url(
+            assignment, historical_assignment
+        )
 
-        If we detect that the new assignment in the LMS has been copied from a
-        historical assignment (perhaps by using the LMS's "course copy" feature)
-        then some of the new assignment's extra values might be taken from the historical assignment.
-        """
-        extra = extra or {}
+        if not document_url:
+            # We can't have a document_url, we shouldnt' try to create an assigmetn yet.
+            return None
 
-        assignment = self.get_assignment(tool_consumer_instance_guid, resource_link_id)
         if not assignment:
-            assignment = Assignment(
-                tool_consumer_instance_guid=tool_consumer_instance_guid,
-                document_url=document_url,
-                resource_link_id=resource_link_id,
+            # We don't have an assignment in the DB but we know to which document url we have to point the current launched one, create a new one
+            # THis might happen for newly launched
+            assignment = self.create_assignment(
+                tool_consumer_instance_guid, resource_link_id
             )
-            self._db.add(assignment)
 
-            # For new assignments check if we are copying
-            # from an existing one on the current launch
-            if historical_assignment := self.get_copied_from_assignment(lti_params):
+            if historical_assignment:
+                # While creating a new assignment we found the assignment we copied this one from
+                # Reference it on the DB
                 assignment.copied_from = historical_assignment
-
-                if historical_assignment.extra.get("group_set_id") and not extra.get(
-                    "group_set_id"
-                ):
-                    extra["group_set_id"] = historical_assignment.extra.get(
+                # And copy over any settings
+                if historical_assignment.extra.get("group_set_id"):
+                    assignment.extra["group_set_id"] = historical_assignment.extra.get(
                         "group_set_id"
                     )
 
+        # Always update the assignment ULR
         assignment.document_url = document_url
+
+        # And metadata based on the launch
         assignment.title = lti_params.get("resource_link_title")
         assignment.description = lti_params.get("resource_link_description")
-        assignment.is_gradable = is_gradable
-        if extra:
-            assignment.extra = extra
+        assignment.is_gradable = self._misc_plugin.is_assignment_gradable(lti_params)
```

Comment:
Have to think about this. In the current state of this PR this are not updated for new and edited assignments, only on launches.